src/appleseed/renderer/meta/tests/test_phasefunction.cpp

```diff
@@ -0,0 +1,337 @@
+
+TEST_SUITE(Renderer_Modeling_PhaseFunction)
+{
+    struct Fixture
+      : public TestFixtureBase
+    {
+        // Number of MC samples to do integrations.
+        static const int NumberOfSamples = 200000;
+        // Number of samples to draw.
+        static const int NumberOfSamplesPlot = 100;
+
+        ParamArray m_base_parameters;
+        const HenyeyPhaseFunctionFactory m_phase_function_factory;
+
+        Fixture()
+          : m_phase_function_factory()
+        {
+            m_base_parameters
+                .insert("scattering", "0.5")
+                .insert("scattering_multiplier", "1.0")
+                .insert("extinction", "0.5")
+                .insert("extinction_coefficient", "1.0");
+        }
+
+        template <typename ReturnType, typename Procedure>
+        ReturnType setup_environment_and_evaluate(Procedure procedure)
+        {
+            TextureStore texture_store(m_scene);
+            TextureCache texture_cache(texture_store);
+
+            boost::shared_ptr<OIIO::TextureSystem> texture_system(
+                OIIO::TextureSystem::create(),
+                boost::bind(&OIIO::TextureSystem::destroy, _1));
+
+            RendererServices renderer_services(
+                m_project,
+                *texture_system);
+
+            boost::shared_ptr<OSL::ShadingSystem> shading_system(
+                new OSL::ShadingSystem(&renderer_services, texture_system.get()));
+
+            Intersector intersector(
+                m_project.get_trace_context(),
+                texture_cache);
+
+            Arena arena;
+            OSLShaderGroupExec sg_exec(*shading_system, arena);
+
+            Tracer tracer(
+                m_scene,
+                intersector,
+                texture_cache,
+                sg_exec);
+
+            ShadingContext shading_context(
+                intersector,
+                tracer,
+                texture_cache,
+                *texture_system,
+                sg_exec,
+                arena,
+                0);
+
+            return procedure(shading_context, arena);
+        }
+
+        // Integrate PDF of phase function (direction sampling) using straightforward Monte-Carlo approach.
+        static float integrate_phase_function_direction_pdf(
+            PhaseFunction& phase_function,
+            ShadingContext& shading_context,
+            Arena& arena)
+        {
+            ShadingRay shading_ray;
+            shading_ray.m_org = Vector3d(0.0f, 0.0f, 0.0f);
+            shading_ray.m_dir = Vector3d(1.0f, 0.0f, 0.0f);
+            void* data = phase_function.evaluate_inputs(shading_context, shading_ray);
+            phase_function.prepare_inputs(arena, shading_ray, data);
+
+            MersenneTwister rng;
+            SamplingContext sampling_context(rng, SamplingContext::RNGMode);
+
+            float integral = 0.0f;
+            for (int i = 0; i < NumberOfSamples; ++i)
+            {
+                sampling_context.split_in_place(2, 1);
```

Comment:
You don't want to split for every sample because that starts a new uncorrelated trajectory. Instead, you want to initialize the sampling context for the correct number of samples:

    SamplingContext sampling_context(rng, SamplingContext::RNGMode, 2, NumberOfSamples);

With that, you can just call `sampling_context.next2<Vector2f>()` in the loop to generate your 2D samples.

You'll need to make `NumberOfSamples` a `size_t` (which is a good practice anyway for counters).